garbage-collection/gc_cleanup.py

```diff
@@ -0,0 +1,148 @@
+# argo garbage collection
+
+import os
+import argparse
+import json, yaml
+from datetime import datetime, tzinfo, timedelta
+import logging
+
+
+from kubernetes import client, config, watch
+from kubernetes.client.rest import ApiException
+from pprint import pprint
+
+
+# config.load_incluster_config()
+
+config.load_kube_config()
+
+api_client = client.ApiClient()
+custom_api = client.CustomObjectsApi(api_client)
+v1_api = client.CoreV1Api(api_client)
+
+def get_pods(workflow,namespace):
+    # check if pods exist in workflow and collect them
+    nodes = list(workflow['status']['nodes'].keys())
+    pods = []
+    for node in nodes:
+        try:
+            api_response = v1_api.read_namespaced_pod_status(node, namespace)
+            pods.append(node)
+        except ApiException as e:
+            # print("Exception when calling CoreV1Api->read_namespaced_pod: %s\n" % e)
+            pass
+    return pods
+
+
+def delete_pods(pod,namespace,body):
+    try:
+        api_response = v1_api.delete_namespaced_pod(pod, namespace, body=body, propagation_policy='Background')
+    except ApiException as e:
+        print("Exception when calling CoreV1Api->delete_namespaced_pod: %s\n" % e)
+    print("{} deleted".format(pod))
+    return api_response
+
+def clean_up(args):
+    # body object for kubernetes api
+    body = client.V1DeleteOptions()
+    # get all workflows
+    try:
+        workflows = custom_api.list_namespaced_custom_object(args.group, args.version, args.namespace, args.plural)
+    except ApiException as e:
+        logging.info("Exception when calling CustomObjectsApi->list_namespaced_custom_object: %s\n" % e)
+
+    # track workflows expired, workflows not expired and pods deleted for logging
+    workflows_expired = []
+    workflows_not_expired = []
+    pods_deleted = []
+    for workflow in workflows['items']:
+        key = workflow['metadata']['name']
+
+        try:
+            finished_at = datetime.strptime(workflow['status']['finishedAt'], '%Y-%m-%dT%H:%M:%SZ')
+        except TypeError:
+            logging.info('could not read workflow {}'.format(key))
+        time_since_completion = (datetime.utcnow() - finished_at).total_seconds()/60/60
```

Comment:
again, if error is raised and caugth, `finished_at` will raise NameError

Reply:
used continue to move to next workflow